src/transpile/cacheblocking.hpp

```diff
@@ -669,30 +648,75 @@ bool CacheBlocking::is_diagonal_op(Operations::Op& op) const
       return true;
     else if(op.name == "t" || op.name == "tdg")
       return true;
-    else if(op.name == "pauli"){
-      is_pauli_str = true;
-    }
   }
   else if(op.type == Operations::OpType::matrix){
-    if (ignore_diagonal_ && Utils::is_diagonal(op.mats[0], .0)){
+    if (Utils::is_diagonal(op.mats[0], .0)){
       return true;
     }
   }
+  else if(op.type == Operations::OpType::diagonal_matrix){
+    return true;
+  }
+  return false;
+}
 
-  if(is_pauli_str){
-    int i;
-    bool is_diag = true;
-    for(i=0;i<op.qubits.size();i++){
-      switch(op.string_params[0][op.qubits.size() - 1 - i]){
-        case 'X':
-        case 'Y':
-          is_diag = false;
-          break;
-        default:
-          break;
+//split 1 pauli gate to 2 pauli gates (inside and outside)
+bool CacheBlocking::split_pauli(const Operations::Op& op,const reg_t blockedQubits,std::vector<Operations::Op>& out,std::vector<Operations::Op>& queue) const
+{
+  reg_t qubits_in_chunk;
+  reg_t qubits_out_chunk;
+  std::string pauli_in_chunk;
+  std::string pauli_out_chunk;
+  int_t i,j,n;
+  bool inside;
+
+  //get inner/outer chunk pauli string
+  n = op.qubits.size();
+  for(i=0;i<n;i++){
+    if(op.string_params[0][n - 1 - i] == 'I')
+      continue;   //remove I
+
+    inside = false;
+    for(j=0;j<blockedQubits.size();j++){
+      if(op.qubits[i] == blockedQubits[j]){
+        inside = true;
+        break;
       }
     }
-    return is_diag;
+    if(inside){
+      qubits_in_chunk.push_back(op.qubits[i]);
+      pauli_in_chunk.push_back(op.string_params[0][n-i-1]);
+    }
+    else{
+      qubits_out_chunk.push_back(op.qubits[i]);
+      pauli_out_chunk.push_back(op.string_params[0][n-i-1]);
+    }
+  }
+
+  if(qubits_out_chunk.size() > 0){  //save in queue
+    std::reverse(pauli_out_chunk.begin(),pauli_out_chunk.end());
+    Operations::Op op;
+    op.type = Operations::OpType::gate;
+    op.name = "pauli";
+    op.qubits = qubits_out_chunk;
+    op.string_params = {pauli_out_chunk};
+    queue.push_back(op);
+  }
+
+  if(qubits_in_chunk.size() > 0){
+    std::reverse(pauli_in_chunk.begin(),pauli_in_chunk.end());
+    //mapping swapped qubits
+    for(i=0;i<qubits_in_chunk.size();i++){
+      qubits_in_chunk[i] = qubitMap_[qubits_in_chunk[i]];
+    }
+
+    Operations::Op op;
+    op.type = Operations::OpType::gate;
+    op.name = "pauli";
+    op.qubits = qubits_in_chunk;
+    op.string_params = {pauli_in_chunk};
```

Comment:
We can add a small `create_pauli_op` or `insert_pauli_op` passing  `{qubits_in_chunk,  qubits_in_chunk}` or `{qubits_out_chunk,  qubits_out_chunk}`.